File: go/problem_solving/leet_code/is_valid/is_valid.go
package is_valid

var pairs = map[rune]rune{
	']': '[',
	')': '(',
	'}': '{',
}

func isValid(s string) bool {
	var stack []rune
	for _, v := range s {
		if match, ok := pairs[v]; ok {
			if len(stack) > 0 && stack[len(stack)-1] == match {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		} else {
			stack = append(stack, v)
		}
	}
	return len(stack) == 0
}

File: ts/problem_solving/leet_code/is_valid/is_valid.ts
export const brackets: Record<string, string> = {
  "}": "{",
  "]": "[",
  ")": "(",
}
export function isValid(s: string): boolean {
  const stack = []
  for (const c of s) {
    if (brackets[c]) {
      if (stack.length > 0 && stack.at(-1) === brackets[c]) {
        stack.pop()
      } else {
        return false
      }
    } else {
      stack.push(c)
    }
  }
  return stack.length === 0
}

console.log(isValid("()")) // true
console.log(isValid("()[]{}")) // true
console.log(isValid("(]")) // false
